design/design-generic-references.md

```diff
@@ -0,0 +1,197 @@
+get and set the values unknown types and once it is done, an update request will
+be issued. In line with cross-resource references, if the field on the referenced
+object, pointed by `toFieldPath`, already has a value then the resolver will
+skip it, meaning it will resolve only once and users will need to delete that
+value for resolver to set it again.
+
+## Alternatives Considered
+
+### Simpler API
+
+```yaml
+spec:
+  externalValues:
+    - group: ec2.aws.crossplane.io
+      version: v1alpha1
+      resource: vpcs
+      name: main-vpc
+      fromFieldPath: spec.forProvider.cidrBlock
+      toFieldPath: spec.cidrBlock
+```
+
+We could certainly have gone with this approach as well but having everything at
+the same level increases the risk of having to break the API to add new features.
+Although it's probably safe that we will not see many new types of sources be
+added, because a lot of use cases are already covered by composition patch types,
+it could be desirable to have multiple source objects to construct a single value
+for example and that'd be a new type. So, we're taking the safer approach of
+letting the API be more open-ended for future developments and pay a small UX
+fee in exchange.
+
+### Multiple Source References
+
+Composition allows information to flow from single resource to another due to
+the nature of how patches are designed. However, there could be cases where you'd
+like to construct a single value using information from multiple resources, like
+an ARN. An API that would allow this could look like the following:
+
+```yaml
+spec:
+  externalValues:
+  - combine:
+    - version: v1
+      resource: configmaps
+      name: common-settings
+      namespace: crossplane-system
+      fieldPath: data.region
+    - version: v1
+      resource: secrets
+      name: accounts
+      namespace: crossplane-system
+      fieldPath: data.prodAWSAccountID
+    fmt: "arn:aws:clouddirectory:%s:%s:schema/published/cognito/1.0"
+    toFieldPath: "spec.forProvider.directoryArn"
+```
+Or alternatively:
+```yaml
+spec:
+  externalValues:
+  - variables:
+    - name: region
+      fromRef:
+        version: v1
+        resource: configmaps
+        name: common-settings
+        namespace: crossplane-system
+        fieldPath: data.region
+    - name: accountId
+      fromRef:
+        version: v1
+        resource: secrets
+        name: accounts
+        namespace: crossplane-system
+        fieldPath: data.prodAWSAccountID
+    template: "arn:aws:clouddirectory:{{ .region }}:{{ .accountId }}:schema/published/cognito/1.0"
+    toFieldPath: "spec.forProvider.directoryArn"
+```
+
+The advantage of this option is that it will allow composite/managed resource to
+get information from multiple resources to construct a single value, which is not
+possible today in any way. However, it's questionable how necessary it is since
+we haven't seen many users reporting that they need this functionality.
+
+## Future Considerations
```

Comment:
One feature that I think could be immediately useful to folks is "obscured references". This would mean that a sensitive value from a `Secret` (or other resource) could be used in a field, but it wouldn't actually be exposed on the resource. I imagine we could just populate the field in the `toFieldPath` with something like `OBSCURED`, but that could be subject to schema type constraints for resources...

Reply:
Hmm yeah makes sense. We're making sensitive fields secret reference but there might be cases where a thing is not sensitive on the surface but is for users. I'll add it.